### PR TITLE
Add identity.xml templating.

### PR DESCRIPTION
**Pull request by hasinidilanka**

### CODE REVIEW

File: modules/distribution/src/repository/resources/conf/default.json [hunk 1]
```diff
@@ -363,7 +363,135 @@
   "identity.events.schemes.OIDCLogoutHandler.properties.enable": false,
   "identity.events.schemes.confirmationCodesInvalidate.subscriptions": [
     "POST_DELETE_USER"
-  ]
+  ],
+
+
+
+  "identity.datasource": "jdbc/WSO2CarbonDB",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: this is not needed, we are using a new database now named WSO2IdentityDB, and we should not be giving the option to change it. (users may still point all to the same DB at TOML)

File: modules/distribution/src/repository/resources/conf/default.json [hunk 2]
```diff
@@ -363,7 +363,135 @@
...
+
+  "identity.issuer_policy": "SelfAndManaged",
+  "identity.token_validation_policy": "CertValidate",
+
+  "OpenID.server_url": "${carbon.protocol}://${carbon.host}:${carbon.management.port}/openidserver",
+  "OpenID.user_pattern": "${carbon.protocol}://${carbon.host}:${carbon.management.port}/openid",
+  "OpenID.login_url": "${carbon.protocol}://${carbon.host}:${carbon.management.port}/authenticationendpoint/openid_login.do",
+  "OpenID.skip_user_consent": false,
+  "OpenID.remember_me_expiry": "7200",
+  "OpenID.disable_dumb_mode": false,
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: let's have these 'openid.**'

File: modules/distribution/src/repository/resources/conf/default.json [hunk 3]
```diff
@@ -363,7 +363,135 @@
...
+  "identity.session_data_persist.temp_data_cleanup.enable": true,
+  "identity.session_data_persist.temp_data_cleanup.pool_size": "20",
+  "identity.session_data_persist.temp_data_cleanup.cleanup_timeout": "40",
+
+  "identity.time_config.session_idle_timeout": "15",
+  "identity.time_config.remember_me_timeout": "20160",
+
+  "identity.security.keystore_directory_path": "${carbon.home}/conf/keystores",
+  "identity.security.key_manager_type": "$conf{keystore.tls.type}",
+  "identity.security.trust_manager_type": "$conf{truststore.type}",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: As discussed offline, let's rename the keys from a product/functionality POV

File: modules/distribution/src/repository/resources/conf/default.json [hunk 4]
```diff
@@ -363,7 +363,135 @@
...
+    "org.wso2.carbon.identity.oauth2.validators.JDBCScopeValidator",
+    "org.wso2.carbon.identity.oauth2.validators.xacml.XACMLScopeValidator"
+  ],
+
+  "OAuth.scope_handlers": [
+    "org.wso2.carbon.identity.oauth2.validators.OIDCScopeHandler"
+  ],
+
+  "OAuth.enable_assertions.UserName": false,
+  "OAuth.enable_access_token_partitioning": false
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: "oauth.**"

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 5]
```diff
@@ -1,79 +1,89 @@
...
+ -->
 
 <Server xmlns="http://wso2.org/projects/carbon/carbon.xml">
 
     <JDBCPersistenceManager>
         <DataSource>
             <!-- Include a data source name (jndiConfigName) from the set of data
                 sources defined in master-datasources.xml -->
-            <Name>{{identity.datasource}}</Name>
+            <Name>jdbc/WSO2CarbonDB</Name>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Review

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 6]
```diff
@@ -331,105 +350,140 @@
...
             all the application keys are saved in to the same table. UserName Assertion
             should be 'true' to use this. -->
-        <EnableAccessTokenPartitioning>false</EnableAccessTokenPartitioning>
+        <EnableAccessTokenPartitioning>{{oauth.token.partitioning.enable}}</EnableAccessTokenPartitioning>
         <!-- user store domain names and mapping to new table name. eg: if you
             provide 'A:foo.com', foo.com should be the user store domain name and 'A'
             represent the relavant mapping of token store table, i.e. tokens will be
             added to a table called IDN_OAUTH2_ACCESS_TOKEN_A. -->
-        <AccessTokenPartitioningDomains><!-- A:foo.com, B:bar.com --></AccessTokenPartitioningDomains>
+        <AccessTokenPartitioningDomains>{{oauth.token.partitioning.domains}}</AccessTokenPartitioningDomains>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: review

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 7]
```diff
@@ -331,105 +350,140 @@
...
-                <SupportedIDTokenEncryptionMethod>A128CBC+HS256</SupportedIDTokenEncryptionMethod>
+                {% for method in oauth.oidc.id_token.supported_encryption_methods %}
+                <SupportedIDTokenEncryptionMethod>{{method}}</SupportedIDTokenEncryptionMethod>
+                {% endfor %}
             </SupportedIDTokenEncryptionMethods>
 
-            <EnableAudiences>true</EnableAudiences>
+            <EnableAudiences>{{oauth.oidc.id_token.enable_audiences}}</EnableAudiences>
             <!-- Comment out to add Audience values to the JWT token (id_token)  -->
-            <!--Audiences>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Audience default value.

File: modules/distribution/src/repository/resources/conf/default.json [hunk 8]
```diff
@@ -363,7 +363,432 @@
...
+  "mgt.recovery.password_recovery.with_notification.enable": false,
+  "mgt.recovery.password_recovery.with_challenge_question.enable": false,
+  "mgt.recovery.password_recovery.with_challenge_question.notify_start": false,
+  "mgt.recovery.password_recovery.with_challenge_question.separator_char": "!",
+  "mgt.recovery.password_recovery.with_challenge_question.min_answers": "2",
+  "mgt.recovery.password_recovery.with_challenge_question.recaptcha.enable": true,
+  "mgt.recovery.password_recovery.with_challenge_question.recaptcha.max_fails": "2",
+  "mgt.recovery.password_recovery.by_admin.offline": false,
+  "mgt.recovery.password_recovery.by_admin.otp": false,
+  "mgt.recovery.password_recovery.by_admin.recovery_link": false,
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: let's keep identity.mgt here

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 (hunk 5 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: jdbc/WSO2IdentityDB

File: modules/distribution/src/repository/resources/conf/default.json [hunk 9]
```diff
@@ -363,7 +363,432 @@
...
+    "Password": "admin"
+  },
+  "scim.support_complex_multivalued": false,
+
+  "scim2.user.endpoint": "${carbon.protocol}://${carbon.host}:${carbon.management.port}/scim2/Users",
+  "scim2.group.endpoint": "${carbon.protocol}://${carbon.host}:${carbon.management.port}/scim2/Groups</GroupEPUrl",
+  "scim2.support_complex_multivalued": true,
+  "scim2.filtering_enhancements_enable": true,
+
+  "mgt.recovery.endpoint.enable": true,
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/default.json]: add identity.

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 10]
```diff
@@ -331,105 +350,140 @@
...
+            <IDTokenCustomClaimsCallBackHandler>{{oauth.oidc.id_token.claim_callback_hander}}</IDTokenCustomClaimsCallBackHandler>
+            <IDTokenExpiration>{{oauth.oidc.id_token.validity_time_s}}</IDTokenExpiration>
+            <UserInfoJWTSignatureAlgorithm>{{oauth.oidc.user_info.jwt_signature_alg}}</UserInfoJWTSignatureAlgorithm>
+            <UserInfoEndpointClaimRetriever>{{oauth.oidc.user_info.endpoint.claim_retriever}}</UserInfoEndpointClaimRetriever>
+            <UserInfoEndpointRequestValidator>{{oauth.oidc.user_info.endpoint.request_validator}}</UserInfoEndpointRequestValidator>
+            <UserInfoEndpointAccessTokenValidator>{{oauth.oidc.user_info.endpoint.token_validator}}</UserInfoEndpointAccessTokenValidator>
+            <UserInfoEndpointResponseBuilder>{{oauth.oidc.user_info.endpoint.response_builder}}</UserInfoEndpointResponseBuilder>
             <SkipUserConsent>false</SkipUserConsent>
             <!-- Sign the ID Token with Service Provider Tenant Private Key-->
             <SignJWTWithSPKey>false</SignJWTWithSPKey>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Infer this

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 11]
```diff
@@ -331,105 +350,140 @@
...
-            <UserInfoEndpointResponseBuilder>org.wso2.carbon.identity.oauth.endpoint.user.impl.UserInfoJSONResponseBuilder</UserInfoEndpointResponseBuilder>
+            <IDTokenIssuerID>{{oauth.oidc.id_token.issuer_id}}</IDTokenIssuerID>
+            <IDTokenCustomClaimsCallBackHandler>{{oauth.oidc.id_token.claim_callback_hander}}</IDTokenCustomClaimsCallBackHandler>
+            <IDTokenExpiration>{{oauth.oidc.id_token.validity_time_s}}</IDTokenExpiration>
+            <UserInfoJWTSignatureAlgorithm>{{oauth.oidc.user_info.jwt_signature_alg}}</UserInfoJWTSignatureAlgorithm>
+            <UserInfoEndpointClaimRetriever>{{oauth.oidc.user_info.endpoint.claim_retriever}}</UserInfoEndpointClaimRetriever>
+            <UserInfoEndpointRequestValidator>{{oauth.oidc.user_info.endpoint.request_validator}}</UserInfoEndpointRequestValidator>
+            <UserInfoEndpointAccessTokenValidator>{{oauth.oidc.user_info.endpoint.token_validator}}</UserInfoEndpointAccessTokenValidator>
+            <UserInfoEndpointResponseBuilder>{{oauth.oidc.user_info.endpoint.response_builder}}</UserInfoEndpointResponseBuilder>
             <SkipUserConsent>false</SkipUserConsent>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Add the negation of prompt_user_consent

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 12]
```diff
@@ -917,79 +1114,111 @@
...
         <Servlet>
-            <Context>/identity/(.*)</Context>
+            {% for servlet in tenant_context.rewrite.servlets %}
+            <Context>{{servlet}}</Context>
+            {% endfor %}
         </Servlet>
     </TenantContextsToRewrite>
-
     <!-- Server Synchronization Tolerance Configuration in seconds -->
     <ClockSkew>300</ClockSkew>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: review

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 13]
```diff
@@ -917,79 +1114,111 @@
     This property restricts federated user association done through UserProfileAdmin admin service.
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: review  [L913](https://github.com/wso2/product-is/pull/4572/files#diff-8439dcc4cbbf52386455006c761214b0L913) 

review [L773](https://github.com/wso2/product-is/pull/4572/files#diff-8439dcc4cbbf52386455006c761214b0L773)

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 14]
```diff
@@ -642,96 +709,147 @@
...
      file recorder. This recorder is disabled by default. Enable it by setting enable="true". To run these recorders,
      EventListener "rg.wso2.carbon.user.mgt.listeners.UserDeletionEventListener" also should be enabled. Which is
      also disabled by default. -->
     <UserDeleteEventRecorders>
-        <UserDeleteEventRecorder name="org.wso2.carbon.user.mgt.recorder.DefaultUserDeletionEventRecorder" enable="false">
+        {% for event_recorder in event.user_delete_event_recorders%}
+        <UserDeleteEventRecorder name="{{event_recorder.name}}" enable="{{event_recorder.enable}}">
             <!-- Un comment below line if you need to write entries to a separate .csv file. Otherwise this will be
             written in to a log file using a separate appender. -->
             <!--<Property name="path">${carbon.home}/repository/logs/delete-records.csv</Property>-->
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: review

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 15]
```diff
@@ -642,96 +709,147 @@
         <ReturnOnlyMappedLocalRoles>false</ReturnOnlyMappedLocalRoles>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: review - where to put in default.json

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 16]
```diff
@@ -1,79 +1,89 @@
...
+        <KeyManagerType>{{key_mgt.key_manager_type}}</KeyManagerType>
+        <TrustManagerType>{{key_mgt.trust_manager_type}}</TrustManagerType>
     </Security>
 
     <Identity>
-        <IssuerPolicy>SelfAndManaged</IssuerPolicy>
-        <TokenValidationPolicy>CertValidate</TokenValidationPolicy>
+        <IssuerPolicy>{{identity.issuer_policy}}</IssuerPolicy>
+        <TokenValidationPolicy>{{identity.token_validation_policy}}</TokenValidationPolicy>
         <BlackList></BlackList>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Review and remove

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 [hunk 17]
```diff
@@ -88,15 +98,15 @@
             https://<HostName>:<MgtTrpProxyPort except 443>/<ProxyContextPath>/<context>
             If above format doesn't satisfy uncomment the following configs and explicitly configure the values
          -->
-        <OpenIDServerUrl>${carbon.protocol}://${carbon.host}:${carbon.management.port}/openidserver</OpenIDServerUrl>
-        <OpenIDUserPattern>${carbon.protocol}://${carbon.host}:${carbon.management.port}/openid</OpenIDUserPattern>
-        <OpenIDLoginUrl>${carbon.protocol}://${carbon.host}:${carbon.management.port}/authenticationendpoint/openid_login.do</OpenIDLoginUrl>
+        <OpenIDServerUrl>{{openid.server_url}}</OpenIDServerUrl>
```
Comment (hasinidilanka) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: No need to template.

File: modules/distribution/src/repository/resources/conf/default.json [hunk 18]
```diff
@@ -367,78 +367,90 @@
...
-  "identity.security.trust_manager_type": "$conf{truststore.type}",
+  "session.data.persist.enable": true,
+  "session.data.persist.temp_data": true,
+  "session.data.persist.db_pool_size": "0",
+  "session.data.cleanup.enable_for": ["expired_data","invalidated_data","temp_data"],
+  "session.data.cleanup.expired_data.expiry_time_mins": "20160",
+  "session.data.cleanup.expired_data.frequency_mins": "1440",
+  "session.data.cleanup.expired_data.in_chunks_of": "50000",
+  "session.data.cleanup.temp_data.db_pool_size": "20",
+  "session.data.cleanup.temp_data.expiry_time_mins": "40",
```
[the remark this answers is not in the document]
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/default.json]: Fixed.

File: modules/distribution/src/repository/resources/conf/default.json (hunk 8 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/default.json]: Fixed.

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 (hunk 10 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Fixed.

File: modules/distribution/src/repository/resources/conf/default.json [hunk 19]
```diff
@@ -363,7 +363,428 @@
   "identity.events.schemes.OIDCLogoutHandler.properties.enable": false,
   "identity.events.schemes.confirmationCodesInvalidate.subscriptions": [
     "POST_DELETE_USER"
-  ]
+  ],
+
+
+
+  "identity.clock_skew": "300",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: let's have this under deployement

File: modules/distribution/src/repository/resources/conf/default.json [hunk 20]
```diff
@@ -363,7 +363,428 @@
...
+
+  "oauth.token.cleanup.enable": true,
+  "oauth.token.cleanup.backup_for_auditing": true,
+  "oauth.token.persist_access_token_alias": true,
+  "oauth.use_entityid_as_issuer_in_oidc_discovery": true,
+  "oauth.authz_code.validity_time_s": "300",
+  "oauth.app_token.validity_time_s": "3600",
+  "oauth.user_token.validity_time_s": "3600",
+  "oauth.refresh_token.validity_time_s": "84600",
+  "oauth.timestamp_skew": "0",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: let's use the same skew we defined earlier, no need to have separate properties

File: modules/distribution/src/repository/resources/conf/default.json [hunk 21]
```diff
@@ -363,7 +363,428 @@
...
+  "identity.token_validation_policy": "CertValidate",
+
+  "oauth.token.cleanup.enable": true,
+  "oauth.token.cleanup.backup_for_auditing": true,
+  "oauth.token.persist_access_token_alias": true,
+  "oauth.use_entityid_as_issuer_in_oidc_discovery": true,
+  "oauth.authz_code.validity_time_s": "300",
+  "oauth.app_token.validity_time_s": "3600",
+  "oauth.user_token.validity_time_s": "3600",
+  "oauth.refresh_token.validity_time_s": "84600",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/default.json]: let's get rid of '_s' suffix from key, and have the unit at value.

Refer : https://github.com/wso2/carbon-kernel/pull/1982

File: modules/distribution/src/repository/resources/conf/default.json (hunk 2 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/default.json]: Fixed.

File: modules/distribution/src/repository/resources/conf/default.json (hunk 4 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/default.json]: Fixed.

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 (hunk 11 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Fixed.

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2 (hunk 17 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/templates/repository/conf/identity/identity.xml.j2]: Fixed.

File: modules/distribution/src/repository/resources/conf/default.json (hunk 19 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/default.json]: Fixed.

File: modules/distribution/src/repository/resources/conf/default.json (hunk 21 above)
Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/default.json]: Fixed.

File: modules/distribution/src/repository/resources/conf/unit-resolve.json [hunk 22]
```diff
@@ -0,0 +1,62 @@
+{
+  "time": {
+
+    "session_data.cleanup.expired_data.expiry_time": "d",
+    "session_data.cleanup.expired_data.frequency": "d",
+    "session_data.cleanup.temp_data.expiry_time": "d",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/unit-resolve.json]: I guess component defaults for these are in minutes, can you please check?

File: modules/distribution/src/repository/resources/conf/unit-resolve.json [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+    "session.timeout.idle_session_timeout": "m",
+    "session.timeout.remember_me_session_timeout": "d",
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/unit-resolve.json]: should be minutes I guess

Comment (hasinidilanka) [reply modules/distribution/src/repository/resources/conf/unit-resolve.json]: Fixed.

**Pull request merged**